internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -1058,3 +1072,69 @@ func (a *Adapter) findSnapshotWithOpenedPR(snapshot *applicationapiv1alpha1.Snap
 	}
 	return nil, nil
 }
+
+// checkAndCancelOldSnapshotsPipelineRun sorts all snapshots for application and cancels all running integrationTest pipelineruns within application
+// removes finalizer before the pipelinerun is set as CancelledRunFinally to be gracefully cancelled
+func (a *Adapter) checkAndCancelOldSnapshotsPipelineRun(application *applicationapiv1alpha1.Application, snapshot *applicationapiv1alpha1.Snapshot) error {
+
+	snapshots, err := a.loader.GetAllSnapshotsForPR(a.context, a.client, application, snapshot.GetLabels()[gitops.PipelineAsCodePullRequestAnnotation], snapshot.GetLabels()[gitops.PipelineAsCodeRepoURLAnnotation])
+	if err != nil {
+		a.logger.Error(err, "Failed to fetch Snapshots for the application",
+			"application.Name:", application.Name)
+		return err
+	}
+	sortedSnapshots := gitops.SortSnapshots(*snapshots)
+	if len(sortedSnapshots) < 2 {
+		return nil
+	}
+
+	for i := 1; i < len(sortedSnapshots); i++ {
+		if gitops.IsComponentSnapshot(&sortedSnapshots[i]) {
+			err = gitops.MarkSnapshotAsCanceled(a.context, a.client, &sortedSnapshots[i], "Snapshot canceled/superseded")
+			if err != nil {
+				a.logger.Error(err, "Failed to mark snapshot as canceled", "snapshot.Name", &sortedSnapshots[i].Name)
+				return err
+			}
+			err = a.cancelAllPipelineRunsForSnapshot(&sortedSnapshots[i])
+		}
+	}
+	return err
+}
+
+// cancelAllPipelineRunsForSnapshot gets all integration test pipelieruns for a given snapshot
+func (a *Adapter) cancelAllPipelineRunsForSnapshot(snapshot *applicationapiv1alpha1.Snapshot) error {
+	// get all integration pipelineruns for a snapshot
+	integrationTestPipelineruns, err := a.loader.GetAllIntegrationPipelineRunsForSnapshot(a.context, a.client, snapshot)
+	if err != nil {
+		a.logger.Error(err, "Failed to get all integration pipelineruns for snapshot", "snapshot.Name", snapshot.Name)
+		return err
+	}
+	if len(integrationTestPipelineruns) < 1 {
+		a.logger.Info("No integrationTest pipelineruns were found for snapshot", "snapshot.Name", snapshot.Name)
+		return nil
+	}
+	for _, plr := range integrationTestPipelineruns {
+		plr := plr
+		pipelinerunStatus, _, err := h.GetRunningIntegrationPipelineRunStatus(&plr)
```

Comment:
The second returned value from `GetRunningIntegrationPipelineRunStatus` are not used here. How about using `!HasPipelineRunFinished(pipelineRun)` directly?

Reply:
Would not we get also pending ones? Or other statuses that we are not interested in?

Reply:
Just feel pending status mean it doesn't finish. We can keep the current. Thanks